### PR TITLE
[Feature] Build custom typeahead component

### PR DESCRIPTION
**Pull request by angeladietz**
## Notion ticket link

[Build custom typeahead component](https://www.notion.so/uwblueprintexecs/Build-custom-typeahead-component-4db55bc30e98458d8a23d56211a40be9)

## Implementation description
* Typeahead component leverages [AsyncTypeahead](https://github.com/ericgio/react-bootstrap-typeahead/)
* Helmet is used to override menu item styling

Demo with github users query:

https://user-images.githubusercontent.com/35577524/137983389-e5565c77-6468-40b2-a733-a43c6e207025.mov

Screen Shot 2021-10-19 at 5 03 49 PM

Screen Shot 2021-10-19 at 5 04 20 PM

## Notes
* Currently used in `permit-holders.tsx` for testing. Will remove all changes to this file before merging
* Main styling difference between design and implementation is that we were unable to include the search icon within the input field due to difficulties with custom rendering the input field

## Checklist
- [x] My PR name is descriptive, is in imperative tense and starts with one of the following: `[Feature]`,`[Improvement]` or `[Fix]`,
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the RCD team on GitHub, or specific people who are associated with this ticket

### CODE REVIEW

File: pages/admin/permit-holders.tsx [hunk 1]
```diff
@@ -118,6 +120,13 @@ type PermitTableInputData = PermitHolder & {
   };
 };
 
+type githubUser = {
```
Comment (OustanDing) [inline pages/admin/permit-holders.tsx:123]: Now that we have the demo GIF, we can remove the sample usage in this file

File: components/Typeahead.tsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+// Typeahead props
+type Props<T extends TypeaheadModel> = Pick<
+  AsyncTypeaheadProps<T>,
+  'isLoading' | 'onSearch' | 'renderMenuItemChildren'
+> & {
+  results: Array<T>;
+  placeholder: string;
+};
+
+export default function Typeahead<T extends TypeaheadModel>(props: Required<Props<T>>) {
```
Comment (OustanDing) [inline components/Typeahead.tsx:22]: Add a JSDoc comment

File: components/Typeahead.tsx [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  const { isLoading, onSearch, renderMenuItemChildren, results, placeholder } = props;
+  const filterBy = () => true;
```
Comment (OustanDing) [inline components/Typeahead.tsx:24]: I think we want to make this a custom prop? What is the function of this `filterBy` function?

File: components/Typeahead.tsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            width: '466px',
+            height: '44px',
+          },
+        }}
+        renderMenuItemChildren={renderMenuItemChildren}
+        renderMenu={(results, menuProps, props) => {
+          menuProps.text = props.text;
+          return (
+            <VStack>
+              <Box></Box>
```
Comment (OustanDing) [inline components/Typeahead.tsx:49]: Remove this?

File: components/Typeahead.tsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+              <Menu {...menuProps}>
+                {results.length == 0 ? (
```
Comment (OustanDing) [inline components/Typeahead.tsx:51]:
```suggestion
                {results.length === 0 ? (
```

File: components/Typeahead.tsx [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                      </Text>
+                    </Center>
+                  )
+                ) : (
+                  results.map((result, index) => (
+                    <>
+                      <MenuItem option={result} position={index}>
+                        {renderMenuItemChildren(result, menuProps, index)}
+                      </MenuItem>
+                      {index != results.length - 1 ? (
```
Comment (OustanDing) [inline components/Typeahead.tsx:71]:
```suggestion
                      {index !== results.length - 1 ? (
```
Good rule of thumb to use strict equality operator (`===`, `!==`)

File: components/Typeahead.tsx [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                        <Divider width="466px" borderWidth="0" color="border.secondary" />
+                      ) : null}
+                    </>
+                  ))
+                )}
+              </Menu>
+            </VStack>
+          );
+        }}
+      ></AsyncTypeahead>
```
Comment (OustanDing) [inline components/Typeahead.tsx:81]:
```suggestion
      />
```

**Review by OustanDing — CHANGES_REQUESTED** [review]

File: components/Typeahead.tsx (hunk 4 above)
Comment (angeladietz) [reply components/Typeahead.tsx:49]: The`Box` adds a small vertical space between the input field and the menu drop down. I changed this to be a `Spacer` instead to make its purpose more evident

**Review by OustanDing — APPROVED** [review]
LGTM! Great work!

**Pull request merged**